### PR TITLE
[SPARK-13369] Make number of consecutive fetch failures for a stage b…

### PR DESCRIPTION
**Pull request by sitalkedia**
## What changes were proposed in this pull request?

The previously hardcoded max 4 retries per stage is not suitable for all cluster configurations. Since spark retries a stage at the sign of the first fetch failure, you can easily end up with many stage retries to discover all the failures. In particular, two scenarios this value should change are (1) if there are more than 4 executors per node; in that case, it may take 4 retries to discover the problem with each executor on the node and (2) during cluster maintenance on large clusters, where multiple machines are serviced at once, but you also cannot afford total cluster downtime. By making this value configurable, cluster managers can tune this value to something more appropriate to their cluster configuration.

unit tests.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Hey, could you actually fill in the fields of the PR description above? The purpose of that template is for you to fill out the parts in parentheses to provide a description of the change.

**Comment by sitalkedia** [comment]
@JoshRosen - Thanks for pointing that out. done!

**Comment by markhamstra** [comment]
I don't understand your reasoning.  If a machine that was holding a needed shuffle file is rebooting, then you should want the Stage to fail so that the results can be recomputed -- if the Worker is gone, then so to are the shuffle files, whether or not they were being served directly or by the shuffle service co-located on the node.  Why would you want to delay that recomputation?

To improve resiliency in the face of worker node failures/reboots, it would be nice to have shuffle files replicated and fetch-able from other worker nodes, but that is quite a different proposition from this PR.

**Comment by sitalkedia** [comment]
@markhamstra Ofcourse we want the stage to fail as soon as we detect fetch failure so the recomputation can be done quickly. But the configuration parameter in discussion here controls the maximum number of times a stage can incur fetch failure before it is aborted (and hence the corresponding job is failed).  Currently the value is hardcode to 4, which means that when we have more than 4 instances of fetch failure due to machine reboot or any other error, we abort the stage and the job fails. This is an issue for our long running Spark jobs which typically see several planned/unplanned machine reboot.

**Comment by sitalkedia** [comment]
Can one of the admins look into this?

**Comment by sitalkedia** [comment]
ping !!

**Comment by srowen** [comment]
@sitalkedia I wouldn't ping 3 times in a few days please.
I guess I'm missing why the number of *consecutive* failures would be expected to increase with the length of the application's runtime?

**Comment by sitalkedia** [comment]
@srowen - sorry for the unnecessary noise. 
Typically in our clusters, we see few machine reboot every hour. We are trying to run Spark jobs which are processing around 100TB of data and runs for several hours. During the run of the job, it very common to see several machine reboot and fetch failures as a result. The longer the job runs, more fetch failure it encounters. That is why we want _the number of consecutive failures_ configurable.

**Comment by markhamstra** [comment]
But if the scheduler behaves the way I expect, then the fetch failures caused by a single node's failure will not produce consecutive failures, but rather fetch failures associated with a single stageAttempt.  Subsequent stageAttempts should end up attempting to fetch the map outputs from a different location than the failed node.  If that is all working as expected, then four consecutive fetch failures would either indicate that four consecutive attempts to fetch the map outputs, each attempt seeking from a different location, all failed through no fault of the Stage's Tasks (very unlikely), or that the Stage's Tasks are causing problems on whichever Executor they run (more likely).

If all that is not working as expected, then increasing MAX_CONSECUTIVE_FETCH_FAILURES would seem to be compensating for broken/unexpected behavior instead of fixing the actual underlying problem -- or I need to improve my understanding and expectations.  Either way, it's probably going to be another week before I get a chance to spend any serious time on this.

**Comment by squito** [comment]
sorry this has sat around so long, I agree this is useful following up on discussion here: https://github.com/apache/spark/pull/17088

I'd reword the description to something more like this:

The previously hardcoded max 4 retries per stage is not suitable for all cluster configurations.  Since spark retries a stage at the sign of the *first* fetch failure, you can easily end up with many stage retries to discover all the failures.  In particular, two scenarios this value should change are (1) if there are more than 4 executors per node; in that case, it may take 4 retries to discover the problem with each executor on the node and (2) during cluster maintenance on large clusters, where multiple machines are serviced at once, but you also cannot afford total cluster downtime.  By making this value configurable, cluster managers can tune this value to something more appropriate to their cluster configuration.

File: core/src/main/scala/org/apache/spark/scheduler/Stage.scala [hunk 1]
```diff
@@ -118,7 +119,7 @@ private[scheduler] abstract class Stage(
    */
   private[scheduler] def failedOnFetchAndShouldAbort(stageAttemptId: Int): Boolean = {
     fetchFailedAttemptIds.add(stageAttemptId)
-    fetchFailedAttemptIds.size >= Stage.MAX_CONSECUTIVE_FETCH_FAILURES
+    fetchFailedAttemptIds.size >= maxConsecutiveFetchFailure
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/Stage.scala]: not a big deal, but since `maxConsecutiveFetchFailure` isn't really changing from stage to stage, so I think rather than making it a member variable of stage, maybe its makes more sense to make it a parameter to this method. thoughts?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -180,6 +180,11 @@ class DAGScheduler(
   /** If enabled, FetchFailed will not cause stage retry, in order to surface the problem. */
   private val disallowStageRetryForTest = sc.getConf.getBoolean("spark.test.noStageRetry", false)
 
+  /** Number of consecutive fetch failures allowed before a stage is aborted */
+  private[scheduler] val maxConsecutiveFetchFailuresPerStage =
+    sc.getConf.getInt("spark.max.fetch.failures.per.stage",
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I don't like this name, it is verbose and it sounds like you might give up on a stage if one *attempt* has too many fetch failures.

I also hate naming though :p

how about "spark.stage.maxAttempts"?  Sort of parallel to "spark.task.maxFailures".

File: docs/configuration.md [hunk 3]
```diff
@@ -1157,6 +1157,13 @@ Apart from these, the following properties are also available, and may be useful
     Should be greater than or equal to 1. Number of allowed retries = this value - 1.
   </td>
 </tr>
+<tr>
+  <td><code>spark.max.fetch.failures.per.stage</code></td>
+  <td>4</td>
+  <td>
+    Number of consecutive fetch failures allowed before a stage is aborted.
```
Comment (squito) [inline docs/configuration.md]: Number of consecutive stage retries allowed before a stage is aborted (since you can have multiple fetch failures in one stage attempt).

**Review by squito — COMMENTED** [review]
minor comments, otherwise I think this is pretty good.

@kayousterhout I recall you not wanting a config for this long ago, but also we didn't really have a reason for putting it in way back then.  Now I think there is a good case for it.  WDYT?

**Comment by sitalkedia** [comment]
@squito - Thanks for taking a look. I am working on resolving the conflicts.

**Comment by kayousterhout** [comment]
GIven discussion with @sitalkedia on other PRs I agree that it makes sense to add now.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 2 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I agree with spark.stage.maxAttempts.  We could also do spark.stage.maxFailures but somehow that seems more easily confused with the number of task failures per stage.

File: core/src/main/scala/org/apache/spark/scheduler/Stage.scala [hunk 4]
```diff
@@ -145,6 +146,6 @@ private[scheduler] abstract class Stage(
 }
 
 private[scheduler] object Stage {
-  // The number of consecutive failures allowed before a stage is aborted
-  val MAX_CONSECUTIVE_FETCH_FAILURES = 4
+  // The number of consecutive fetch failures allowed before a stage is aborted
+  val DEFAULT_MAX_CONSECUTIVE_FETCH_FAILURES = 4
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/Stage.scala]: Given that this is used in the DAGScheduler, not here, I think it makes more sense to put it in the DAGScheduler object.

File: core/src/main/scala/org/apache/spark/scheduler/Stage.scala [hunk 5]
```diff
@@ -58,6 +58,7 @@ private[scheduler] abstract class Stage(
     val numTasks: Int,
     val parents: List[Stage],
     val firstJobId: Int,
+    val maxConsecutiveFetchFailure: Int,
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/Stage.scala]: nit: this should have an "s" at the end (same elsewhere)

File: core/src/main/scala/org/apache/spark/scheduler/Stage.scala (hunk 1 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/Stage.scala]: What about instead cleaning this code path up a bit (this isn't the fault of your PR, but your PR could help make things better here!) by:

(1) eliminate this method
(2) make the fetchFailedAttemptIds variable non-private
(3) In the DAGScheduler on line 1268 (right after the else), add "stage.fetchFailedAttemptIds.add(task.stageAttemptId"
(4) change the call to this method to instead directly check "(stage.fetchFailedAttemptIds.size >= maxConsec..."

IMO as is, this function is hurting understandability as-is.  This change would also allow you to eliminate maxConsecutiveFetchFailure from the Stage object altogether.

**Comment by sitalkedia** [comment]
@squito, @kayousterhout - Sorry, I accidentally deleted the old branch. Created a new PR https://github.com/apache/spark/pull/17307 with suggested change.
